### PR TITLE
DAOS-3610 dmg: Add system list-pools command

### PR DESCRIPTION
**Pull request by kjacque**
This patch adds a dmg command to list all pools in a
DAOS system. The output format matches what is currently
available in the legacy dmg_old tool.

- Add unit tests for MgmtSvc.ListPools
- Add unit tests for List Pools dRPC handler.
- Implement Go client interface for List Pools.
- Add DMG user interface for the new command.
- Simplify dRPC handler and Protobuf definitions.
- Update documentation.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

**Comment by kjacque** [comment]
Sample output:
```
kris@ratty:~$ dmg -i system list-pools
localhost:10001: connected
List-Pools command succeeded
No pools in system
kris@ratty:~$ dmg -i pool create -s 1G
localhost:10001: connected
Creating DAOS pool with 1GB SCM and 0B NvMe storage (1.000 ratio)
Pool-create command SUCCEEDED: UUID: 88f3ba9f-62ff-494a-a9f8-6d81485fc630, Service replicas: 0
kris@ratty:~$ dmg -i system list-pools
localhost:10001: connected
List-Pools command succeeded
88f3ba9f-62ff-494a-a9f8-6d81485fc630 0
```

File: src/control/cmd/dmg/system.go [hunk 1]
```diff
@@ -82,3 +86,47 @@ func (cmd *systemMemberQueryCmd) Execute(args []string) error {
...
+		if len(pool.SvcReplicas) > 0 {
+			b.WriteString(" ")
+			for i, rep := range pool.SvcReplicas {
+				if i != 0 {
+					b.WriteString(":")
+				}
+				fmt.Fprintf(&b, "%d", rep)
+			}
+		}
+		b.WriteString("\n")
```
Comment (mjmac) [inline src/control/cmd/dmg/system.go]: This output is a little hard to parse/understand. Maybe incorporate tabwriter as @tanabarr is doing in his PR to improve storage scan output? That will allow you to present a nice table-based display with column names. e.g. https://github.com/daos-stack/daos/pull/1495/files#diff-bd69096e8b80741d25d70e29f5d87a6eR124

File: src/control/cmd/dmg/system.go [hunk 2]
```diff
@@ -82,3 +86,47 @@ func (cmd *systemMemberQueryCmd) Execute(args []string) error {
...
+		return errors.New("No configuration loaded")
+	}
+	req := client.ListPoolsReq{SysName: cmd.config.SystemName}
+	resp, err := cmd.conns.ListPools(req)
+	if err != nil {
+		cmd.log.Infof("List-Pools command failed: %s\n", err.Error())
+		return err
+	}
+
+	cmd.log.Info("List-Pools command succeeded\n")
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/cmd/dmg/system.go]: I added this entirely for consistency with what we're doing elsewhere. I don't really like it either, so I'm happy to remove it.

File: src/control/cmd/dmg/system.go [hunk 3]
```diff
@@ -82,3 +86,47 @@ func (cmd *systemMemberQueryCmd) Execute(args []string) error {
...
+
+// Execute is run when systemListPoolsCmd activates
+func (cmd *systemListPoolsCmd) Execute(args []string) error {
+	if cmd.config == nil {
+		return errors.New("No configuration loaded")
+	}
+	req := client.ListPoolsReq{SysName: cmd.config.SystemName}
+	resp, err := cmd.conns.ListPools(req)
+	if err != nil {
+		cmd.log.Infof("List-Pools command failed: %s\n", err.Error())
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/cmd/dmg/system.go]: Makes sense to me. We'll want to go through other dmg commands to make this consistent - I think we are just printing the error in a number of places.

File: src/control/cmd/dmg/system.go (hunk 1 above)
Comment (kjacque) [reply src/control/cmd/dmg/system.go]: I was going to ask about this actually - do we want some standardized outputter in DMG that accepts, say, a list of dictionaries with the column names as keys, and displays them in properly tabbed columns? If we find ourselves repeating the same display code in multiple places.

Comment (mjmac) [reply src/control/cmd/dmg/system.go]: That's a great idea... Does it make sense to do that as part of this PR? I'd support that because sometimes it's hard to develop that sort of thing unless you have a specific use case in mind, but I wouldn't insist on it for landing this one.

File: src/control/cmd/dmg/system.go [hunk 4]
```diff
@@ -82,3 +86,47 @@ func (cmd *systemMemberQueryCmd) Execute(args []string) error {
...
+	}
+
+	cmd.log.Info("List-Pools command succeeded\n")
+	if len(resp.Pools) == 0 {
+		cmd.log.Info("No pools in system\n")
+		return nil
+	}
+
+	for _, pool := range resp.Pools {
+		var b strings.Builder
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/cmd/dmg/system.go]: Fixed.

File: src/control/cmd/dmg/system.go (hunk 1 above)
Comment (kjacque) [reply src/control/cmd/dmg/system.go]: I'll look into it and see what it takes. It's the kind of thing I'd normally break into another PR, but if it's not big, it may be worth it to add it now.

File: src/mgmt/tests/mocks.c [hunk 5]
```diff
@@ -168,6 +168,106 @@ mock_ds_mgmt_pool_delete_acl_teardown(void)
...
+		ds_mgmt_list_pools_npools = *npools;
+
+	ds_mgmt_list_pools_poolsp_ptr = (void *)poolsp;
+	if (poolsp != NULL && ds_mgmt_list_pools_poolsp_out != NULL) {
+		size_t	pool_array_size;
+
+		pool_array_size = ds_mgmt_list_pools_len_out *
+				  sizeof(struct mgmt_list_pools_one);
+
+		D_ALLOC(*poolsp, pool_array_size);
```
Comment (kccain) [inline src/mgmt/tests/mocks.c]: can this be reduced to a D_ALLOC_ARRAY() without need to compute pool_array_size?

File: src/mgmt/tests/mocks.c [hunk 6]
```diff
@@ -168,6 +168,106 @@ mock_ds_mgmt_pool_delete_acl_teardown(void)
...
 }
 
+int				ds_mgmt_list_pools_return;
+char				ds_mgmt_list_pools_group[DAOS_SYS_NAME_MAX];
+void				*ds_mgmt_list_pools_npools_ptr;
+uint64_t			ds_mgmt_list_pools_npools;
+void				*ds_mgmt_list_pools_poolsp_ptr;
+struct mgmt_list_pools_one	*ds_mgmt_list_pools_poolsp_out;
+void				*ds_mgmt_list_pools_len_ptr;
+size_t				ds_mgmt_list_pools_len_out;
```
Comment (kccain) [inline src/mgmt/tests/mocks.c]: nit: indented one tab too far?

File: src/mgmt/tests/mocks.h [hunk 7]
```diff
@@ -78,4 +79,20 @@ extern void		*ds_mgmt_pool_delete_acl_result_ptr;
 void mock_ds_mgmt_pool_delete_acl_setup(void);
 void mock_ds_mgmt_pool_delete_acl_teardown(void);
 
+/*
+ * Mock ds_mgmt_list_pools
+ */
+extern int				ds_mgmt_list_pools_return;
+extern char				ds_mgmt_list_pools_group[];
+extern void				*ds_mgmt_list_pools_npools_ptr;
+extern uint64_t				ds_mgmt_list_pools_npools;
```
Comment (kccain) [inline src/mgmt/tests/mocks.h]: nit: indented one tab too far?

File: src/control/cmd/dmg/system.go (hunk 1 above)
Comment (kjacque) [reply src/control/cmd/dmg/system.go]: I went ahead and added a formatter. Based it on the logic in Tom's patch.

File: doc/admin/pool_operations.md [hunk 8]
```diff
@@ -55,6 +55,12 @@ two pool service replica on rank 0 and 1.
 $ dmg pool destroy --pool=${puuid}
 ```
 
+**To see a list of the pools in your DAOS system:**
+
+```
+$ dmg system list-pools
```
[the remark this answers is not in the document]
Comment (kjacque) [reply doc/admin/pool_operations.md]: In the sample output for the daos tool, it looked like these were colon-separated. Do we want them to be comma-separated?

File: src/proto/mgmt/pool.proto [hunk 9]
```diff
@@ -58,17 +58,17 @@ message PoolDestroyResp {
...
 	string sys = 1; // DAOS system identifier
-	uint64 numPools = 2; // Client response buffer capacity in number of pools
 }
 
+// ListPoolsResp returns the list of pools in the system.
 message ListPoolsResp {
 	message Pool {
 		string uuid = 1; // uuid of pool
-		string svcreps = 2; // pool service replicas, comma-separated integers
+		repeated int32 svcreps = 2; // pool service replica ranks
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/proto/mgmt/pool.proto]: I think the formatting should be done as close to the display layer as possible. I think we should probably change PoolCreateResp to use an array of ints, personally. Wondering if @mjmac has an opinion on this.

File: doc/admin/pool_operations.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kccain) [reply doc/admin/pool_operations.md]: We want them to be comma-separated. The dmg_old --help shows the colon-separated syntax. But when the tools proposal was drafted it was decided to use comma-separated. daos --help (in the --svc argument documentation) shows a comma-separated example.

File: src/proto/mgmt/pool.proto [hunk 10]
```diff
@@ -58,17 +58,17 @@ message PoolDestroyResp {
...
 	}
 	int32 status = 1; // DAOS error code
-	repeated Pool pools = 2; // pools list (max length ListPoolsReq.numPools)
-	uint64 numPools = 3; // number of pools in system
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/proto/mgmt/pool.proto]: In Go we don't preallocate a buffer for the Protobuf response, so the whole concept of requesting the length, allocating a buffer, and then sending a separate RPC to request the content doesn't really make sense.

Things could get nasty in the dRPC protocol itself - we have a message max size so it may become necessary to chunk content at that point if it's too big. But I'm not sure how knowing the total number of pools or containers would fit into that - we'd want to be able to ask for a specific subset at a time. Not sure how we should approach that, but I think it's good to start thinking about.

File: src/control/cmd/dmg/system.go [hunk 11]
```diff
@@ -82,3 +86,47 @@ func (cmd *systemMemberQueryCmd) Execute(args []string) error {
...
+	}
+
+	var b strings.Builder
+	for _, pool := range resp.Pools {
+		b.WriteString(pool.UUID)
+		if len(pool.SvcReplicas) > 0 {
+			b.WriteString(" ")
+			for i, rep := range pool.SvcReplicas {
+				if i != 0 {
+					b.WriteString(":")
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/cmd/dmg/system.go]: Ah, I was looking at this page as reference: https://wiki.hpdd.intel.com/display/DAOS/Pool+and+Container+Listing+Design

I will switch to commas.

File: doc/admin/pool_operations.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kjacque) [reply doc/admin/pool_operations.md]: Ah, just got to your comment below. Will add some more details. I've also revised the output to be a bit more human-friendly.

File: src/control/cmd/dmg/system.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (kccain) [reply src/control/cmd/dmg/system.go]: I just updated that page with a heads up about the transition. Thanks for pointing out that issue.
Also I did review the output of dmg pool create and confirmed it does report svcreps output that is comma-separated integers.

File: src/mgmt/tests/mocks.c (hunk 5 above)
Comment (kjacque) [reply src/mgmt/tests/mocks.c]: Good catch, this was lingering from one of my earlier implementation attempts.

File: src/mgmt/tests/mocks.c (hunk 6 above)
Comment (kjacque) [reply src/mgmt/tests/mocks.c]: Checking the whitespace in my IDE, it looks OK, tabs are correct width. We've seen this a few times, not sure what's causing Github to display it differently.

File: src/mgmt/tests/mocks.h (hunk 7 above)
Comment (kjacque) [reply src/mgmt/tests/mocks.h]: Same as above - in my IDE the tabs are set up correctly.

File: src/proto/mgmt/pool.proto (hunk 9 above)
[the remark this answers is not in the document]
Comment (mjmac) [reply src/proto/mgmt/pool.proto]: I agree with the inclination toward keeping the data layer about data and the display layer about display. This seems like a good change to me.

File: doc/admin/pool_operations.md [hunk 12]
```diff
@@ -61,6 +61,19 @@ $ dmg pool destroy --pool=${puuid}
...
+replicas. For example:
+
+```
+$ dmg system list-pools
+localhost:10001: connected
+Pool UUID					Svc Replicas
+---------					------------
+2a8ec3b2-729b-4617-bf51-77f37f764194	0,1
+a106d667-5c5d-4d6f-ac3a-89099196c41a	0
+85141a07-e3ba-42a6-81c2-3f18253c5e47	0
```
Comment (mjmac) [inline doc/admin/pool_operations.md]: This looks good, although there seems to be an alignment issue even when viewed in rich diff mode. Is there any other information we can display here? e.g. pool used/free

**Comment by kjacque** [comment]
> Maybe something to handle in a follow up with more planning -- I understand the desire to get a MVP going, but is there any opportunity to filter these results on the server side? Just thinking about how this will be used by an admin -- I can imagine wanting to filter based on certain criteria (e.g. pool capacity above/below a certain %, by owner, etc).

I'm not sure what the future plans are for expanding the pool listing capabilities. @kccain may have a better idea.

File: doc/admin/pool_operations.md (hunk 12 above)
Comment (kjacque) [reply doc/admin/pool_operations.md]: Whoops. I had "fixed" it because it looked weird when I pasted it into my editor, the actual output is tabbed correctly in my terminal. Copying it directly this time.

File: src/control/client/types.go [hunk 13]
```diff
@@ -469,3 +469,28 @@ func (acl *AccessControlList) Empty() bool {
...
+
+// PoolDiscovery represents the basic discovery information for a pool.
+type PoolDiscovery struct {
+	UUID        string // Unique identifier
+	SvcReplicas []int  // Ranks of pool service replicas
+}
+
+// poolDiscoveriesFromPB converts the protobuf ListPoolsResp_Pool structures to
+// PoolDiscovery structures.
+func poolDiscoveriesFromPB(pbPools []*mgmtpb.ListPoolsResp_Pool) []*PoolDiscovery {
```
Comment (tanabarr) [inline src/control/client/types.go]: @mjmac is now doing this slightly differently from https://github.com/daos-stack/daos/pull/1480
we probably want to be consistent (although could be done as a follow-up as there may be other areas that need migrating)

File: src/proto/mgmt/pool.proto (hunk 9 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/proto/mgmt/pool.proto]: agreed that we should change PoolCreateResp in line with this approach
